packages/astro/src/transitions/router.ts

```diff
@@ -453,9 +453,11 @@ async function transition(
 	if (navigationType !== 'traverse') {
 		updateScrollPosition({ scrollX, scrollY });
 	}
-	if (samePage(from, to) && !!to.hash) {
-		moveToLocation(to, from, options, document.title, historyState);
-		return;
+	if (samePage(from, to)) {
+		if ((direction === 'forward' && !!to.hash) || (direction === 'back' && !!from.hash)) {
```

Comment:
```suggestion
		if ((direction !== 'back' && to.hash) || (direction === 'back' && from.hash)) {
```

Reply:
When changing `event.direction` in a listener, there might exist other values with the same meaning as `forward`